ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -283,38 +283,114 @@ private String formatQualifiedName(QualifiedName name) {
       return new Pair<>("(!" + exprString + ")", Schema.BOOLEAN_SCHEMA);
     }
 
+    private String nullCheckPrefix(ComparisonExpression.Type type) {
+      switch(type) {
+        case IS_DISTINCT_FROM:
+          return "(((Object)%1$s) == null || ((Object)%2$s) == null) ? "
+              + "((((Object)%1$s) == null ) ^ (((Object)%2$s) == null )) : ";
+        default:
+          return "(((Object)%1$s) == null || ((Object)%2$s) == null) ? false : ";
+      }
+    }
+
+    private String visitStringComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "(%1$s.equals(%2$s))";
+        case NOT_EQUAL:
+        case IS_DISTINCT_FROM:
+          return "(!%1$s.equals(%2$s))";
+        case GREATER_THAN_OR_EQUAL:
+        case GREATER_THAN:
+        case LESS_THAN_OR_EQUAL:
+        case LESS_THAN:
+          return "(%1$s.compareTo(%2$s) " + type.getValue() + " 0)";
+        default:
+          throw new KsqlException("Unexpected string comparison: " + type.getValue());
+      }
+    }
+
+    private String visitArrayComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "(java.util.Arrays.equals(%1$s, %2$s))";
+        case NOT_EQUAL:
+        case IS_DISTINCT_FROM:
+          return "!(java.util.Arrays.equals(%1$s, %2$s))";
+        default:
+          throw new KsqlException("Unexpected array comparison: " + type.getValue());
+      }
+    }
+
+    private String visitMapComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "(%1$s.equals(%2$s))";
+        case NOT_EQUAL:
+        case IS_DISTINCT_FROM:
+          return " (!%1$s.equals(%2$s))";
+        default:
+          throw new KsqlException("Unexpected map comparison: " + type.getValue());
+      }
+    }
+
+    private String visitScalarComparisonExpression(ComparisonExpression.Type type) {
+      switch (type) {
+        case EQUAL:
+          return "((%1$s <= %2$s) && (%1$s >= %2$s))";
```

Comment:
Are we happy these types of comparisons work for different types, e.g. BIGINT and INT, where the larger of the two would overflow the lesser?

(Only raising as I'm not sure how these are implemented under the hood, so don't know if they use standard Java scalar types and promotions, or custom code).  

Might be worth adding tests to explicitly test this.

Reply:
Is the concern that they may not work or that they should not work? I think they should work, and they do work - the test for these exercises case.